src/persist/src/operators/stream.rs

```diff
@@ -41,46 +51,252 @@ where
-            move |((k, v), ts, diff)| -> Result<((K, V), u64, isize), (String, u64, isize)> {
-                let res = write.write(&[((k.clone(), v.clone()), ts, diff)]);
-                // TODO: Don't do this synchronously.
-                res.recv().map_err(|err| {
-                    let err_str = format!("persisting data: {}", err);
-                    (err_str, ts, 1)
-                })?;
-                Ok(((k, v), ts, diff))
-            },
-        );
-
-        // Replay the previously persisted data, if any.
-        let (ok_previous, err_previous) = match read.snapshot() {
-            Err(err) => (
-                operator::empty(&self.scope()),
-                // TODO: Figure out how to make these retractable.
-                vec![(format!("replaying persisted data: {}", err), 0, 1)]
-                    .to_stream(&mut self.scope()),
-            ),
-            Ok(snapshot) => operators::replay(&mut self.scope(), snapshot),
-        };
-
-        let ok_all = ok_previous.concat(&ok_new);
-        let err_all = err_previous.concat(&err_new);
-        (ok_all, err_all)
+        let operator_name = format!("persist({})", name);
+        let mut persist_op = OperatorBuilder::new(operator_name.clone(), self.scope());
+
+        let mut input = persist_op.new_input(&self, Pipeline);
+
+        let (mut data_output, data_output_stream) = persist_op.new_output();
+        let (mut error_output, error_output_stream) = persist_op.new_output();
+
+        let mut buffer = Vec::new();
+        let mut write_futures = HashMap::new();
+        let mut input_frontier =
+            Antichain::from_elem(<G::Timestamp as timely::progress::Timestamp>::minimum());
+
+        persist_op.build(move |_capabilities| {
+            move |frontiers| {
+                let mut data_output = data_output.activate();
+                let mut error_output = error_output.activate();
+
+                // Write out everything and forward, keeping the write futures.
+                input.for_each(|cap, data| {
+                    data.swap(&mut buffer);
+
+                    let write_future = write.write(buffer.iter().as_ref());
+
+                    let mut session = data_output.session(&cap);
+                    session.give_vec(&mut buffer);
+
+                    let write_futures = &mut write_futures
+                        .entry(cap.retain_for_output(1))
+                        .or_insert_with(|| Vec::new());
+                    write_futures.push(write_future);
+                });
+
+                // Block on outstanding writes when the input frontier advances.
+                // This way, when the downstream frontier advances, we know that all writes that
+                // are before it are done.
+                let new_input_frontier = frontiers[0].frontier();
+                let progress =
+                    !PartialOrder::less_equal(&new_input_frontier, &input_frontier.borrow());
+
+                if progress {
```

Comment:
nit: can we invert the control flow here to make this `if progress { return }`? dunno about others, but I personally find that style easier to reason about